File: docs/basics/caching/05-failures/java/src/ResilientCache.java
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;

public class ResilientCache<K, V> {
    private final Map<K, V> storage = new HashMap<>();
    private final int maxFailures;
    private final AtomicInteger failureCount = new AtomicInteger(0);
    private volatile boolean circuitOpen = false;

    public ResilientCache(int maxFailures) {
        this.maxFailures = maxFailures;
    }

    public V get(K key) {
        if (circuitOpen) {
            throw new RuntimeException("Circuit breaker open - cache unavailable");
        }
        return storage.get(key);
    }

    public void put(K key, V value) {
        if (circuitOpen) {
            throw new RuntimeException("Circuit breaker open - cache unavailable");
        }
        storage.put(key, value);
    }

    // Circuit breaker pattern - open circuit after too many failures
    public V getWithCircuitBreaker(K key, Supplier<V> fallback) {
        if (circuitOpen) {
            return fallback.get(); // Graceful degradation
        }

        try {
            V result = storage.get(key);
            if (result != null) {
                return result; // Success - reset failure count
            }
            failureCount.set(0);
            return fallback.get();
        } catch (Exception e) {
            if (failureCount.incrementAndGet() >= maxFailures) {
                circuitOpen = true;
            }
            throw e;
        }
    }

    // Cache warming - pre-populate cache to avoid cold starts
    public void warm(Supplier<Map<K, V>> dataLoader) {
        try {
            Map<K, V> warmData = dataLoader.get();
            storage.putAll(warmData);
        } catch (Exception e) {
            // Log warning but don't fail - cache can still serve requests
            System.err.println("Cache warming failed: " + e.getMessage());
        }
    }

    // Reset circuit breaker (for testing/admin purposes)
    public void reset() {
        circuitOpen = false;
        failureCount.set(0);
    }

    public boolean isCircuitOpen() {
        return circuitOpen;
    }

    public int getFailureCount() {
        return failureCount.get();
    }
}

File: docs/system-design/01-fundamentals/caching/05-failures/python/src/resilient_cache.py
from typing import Dict, TypeVar, Generic, Callable, Any
import threading

K = TypeVar('K')
V = TypeVar('V')


class ResilientCache(Generic[K, V]):
    def __init__(self, max_failures: int = 3):
        self._storage: Dict[K, V] = {}
        self._max_failures = max_failures
        self._failure_count = 0
        self._circuit_open = False
        self._lock = threading.Lock()

    def get(self, key: K) -> V:
        if self._circuit_open:
            raise RuntimeError("Circuit breaker open - cache unavailable")
        return self._storage.get(key)

    def put(self, key: K, value: V) -> None:
        if self._circuit_open:
            raise RuntimeError("Circuit breaker open - cache unavailable")
        self._storage[key] = value

    def get_with_circuit_breaker(self, key: K, fallback: Callable[[], V]) -> V:
        """Circuit breaker pattern with graceful degradation"""
        if self._circuit_open:
            try:
                return fallback()  # Graceful degradation
            except Exception as e:
                # Even in circuit open state, if fallback fails, we might want to track this
                # but for now, just re-raise
                raise e

        try:
            result = self._storage.get(key)
            if result is not None:
                with self._lock:
                    self._failure_count = 0  # Reset on success
                return result

            # Key not in cache, try fallback
            result = fallback()
            with self._lock:
                self._failure_count = 0  # Reset on success
            return result

        except Exception as e:
            with self._lock:
                self._failure_count += 1
                if self._failure_count >= self._max_failures:
                    self._circuit_open = True
            raise e

    def warm(self, data_loader: Callable[[], Dict[K, V]]) -> None:
        """Cache warming - pre-populate cache to avoid cold starts"""
        try:
            warm_data = data_loader()
            self._storage.update(warm_data)
        except Exception as e:
            # Log warning but don't fail - cache can still serve requests
            print(f"Cache warming failed: {e}")

    def reset(self) -> None:
        """Reset circuit breaker (for testing/admin purposes)"""
        with self._lock:
            self._circuit_open = False
            self._failure_count = 0

    @property
    def is_circuit_open(self) -> bool:
        return self._circuit_open

    @property
    def failure_count(self) -> int:
        return self._failure_count
